qiskit_dynamics/pulse/pulse_to_signals.py

```diff
@@ -254,3 +264,59 @@ def _get_channel(self, channel_name: str):
             raise QiskitError(
                 f"Invalid channel name {channel_name} given to {self.__class__.__name__}."
             ) from error
+
+
+def get_samples(pulse: SymbolicPulse):
+    """Return samples filled according to the formula that the pulse
+    represents and the parameter values it contains.
+
+    Args:
+        pulse: SymbolicPulse class.
+    Returns:
+        Samples of the pulse.
+    Raises:
+        PulseError: When parameters are not assigned.
+        PulseError: When expression for pulse envelope is not assigned.
+        PulseError: When a free symbol value is not defined in the pulse instance parameters.
+    """
+    envelope = pulse.envelope
+    pulse_params = pulse.parameters
+    if pulse.is_parameterized():
+        raise PulseError("Unassigned parameter exists. All parameters must be assigned.")
+
+    if envelope is None:
+        raise PulseError("Pulse envelope expression is not assigned.")
+
+    args = []
+    for symbol in sorted(envelope.free_symbols, key=lambda s: s.name):
+        if symbol.name == "t":
+            times = Array(np.arange(0, pulse_params["duration"]) + 1 / 2)
+            args.insert(0, times.data)
+            continue
+        try:
+            args.append(pulse_params[symbol.name])
+        except KeyError as ex:
+            raise PulseError(
+                f"Pulse parameter '{symbol.name}' is not defined for this instance. "
+                "Please check your waveform expression is correct."
+            ) from ex
+    return _lru_cache_expr(envelope, Array.default_backend())(*args)
+
+
+@functools.lru_cache(maxsize=None)
+def _lru_cache_expr(expr: sym.Expr, backend):
```

Comment:
Add type hint for `backend` as well as the return type.

Reply:
added at https://github.com/Qiskit/qiskit-dynamics/pull/149/commits/4dab802b8e400282a022df72341da5aff353b5b0